File: 079_word_search/Solution.py
class Solution:
    def exist(self, board: list, word: str) -> bool:
        m, n = len(board), len(board[0])
        word_len = len(word)
        directions = [[-1, 0], [0, -1], [1, 0], [0, 1]]
        visited = [[False] * n for _ in range(m)]

        def dfs(i: int, j: int, str_idx: int) -> bool:
            if str_idx == word_len - 1: return board[i][j] == word[str_idx]

            if board[i][j] != word[str_idx]: return False

            visited[i][j] = True
            for direction in directions:
                new_i = i + direction[0]
                new_j = j + direction[1]
                if in_area(new_i, new_j) and not visited[new_i][new_j]:
                    if dfs(new_i, new_j, str_idx + 1):
                        return True
            visited[i][j] = False

            return False

        def in_area(i: int, j: int):
            return 0 <= i < m and 0 <= j < n

        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0): return True

        return False


if __name__ == '__main__':
    board = [
        ['A','B','C','E'],
        ['S','F','C','S'],
        ['A','D','E','E']
    ]

    solution = Solution()
    print(solution.exist(board, 'ABCCED'))
    print(solution.exist(board, 'SEE'))
    print(solution.exist(board, 'ABCB'))

File: 079_word_search/Solution.java
class Solution {
    private int m, n;
    private boolean[][] visited;
    private final int[][] directions = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
    private char[][] board;
    private String word;
    private int wordLen;

    public boolean exist(char[][] board, String word) {
        if (board == null || board.length == 0 || board[0].length == 0) return false;
        m = board.length;
        n = board[0].length;
        this.visited = new boolean[m][n];
        this.board = board;
        this.word = word;
        this.wordLen = word.length();

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(i, j, 0)) return true;
            }
        }

        return false;
    }

    private boolean dfs(int i, int j, int strIdx) {
        if (strIdx == wordLen - 1) return board[i][j] == word.charAt(strIdx);

        if (board[i][j] != word.charAt(strIdx)) return false;

        visited[i][j] = true;
        for (int k = 0; k < 4; k++) {
            int newI = i + directions[k][0];
            int newJ = j + directions[k][1];
            if (inArea(newI, newJ) && !visited[newI][newJ]) {
                if (dfs(newI, newJ, strIdx + 1)) return true;
            }
        }
        visited[i][j] = false;

        return false;
    }

    private boolean inArea(int i, int j) {
        return 0 <= i && i < m && 0 <= j && j < n;
    }

    public static void main(String[] args) {
        char[][] board = {
            {'A', 'B', 'C', 'E'},
            {'S', 'F', 'C', 'S'},
            {'A', 'D', 'E', 'E'}
        };

        Solution solution = new Solution();
        System.out.println(solution.exist(board, "ABCCE"));
        System.out.println(solution.exist(board, "SEE"));
        System.out.println(solution.exist(board, "ABCB"));
    }
}
